test/extensions/filters/network/dubbo_proxy/route_matcher_test.cc

```diff
@@ -914,6 +914,298 @@ name: config
+  const std::string yaml = R"EOF(
+name: config
+interface: org.apache.dubbo.demo.DemoService
+routes:
+  - match:
+      method:
+        name:
+          exact: "method1"
+    route:
+      metadata_match:
+        filter_metadata:
+          envoy.lb:
+            k1: v1
+            k2: v2
+      weighted_clusters:
+        clusters:
+          - name: cluster1
+            weight: 2000
+            metadata_match:
+              filter_metadata:
+                envoy.lb:
+                  k3: v3
+          - name: cluster2
+            weight: 3000
+          - name: cluster3
+            weight: 5000
+            metadata_match:
+              filter_metadata:
+                envoy.lb:
+                  k2: v3
+)EOF";
+
+  envoy::extensions::filters::network::dubbo_proxy::v3::RouteConfiguration config =
+      parseRouteConfigurationFromV2Yaml(yaml);
+  auto invo = std::make_shared<RpcInvocationImpl>();
+  MessageMetadata metadata;
+  metadata.setInvocationInfo(invo);
+  invo->setServiceName("org.apache.dubbo.demo.DemoService");
+  invo->setMethodName("method1");
+
+  setDefaultInvocationCallback(*invo);
+  NiceMock<Server::Configuration::MockServerFactoryContext> context;
+  SingleRouteMatcherImpl matcher(config, context);
+
+  ProtobufWkt::Value v1, v2, v3;
+  v1.set_string_value("v1");
+  v2.set_string_value("v2");
+  v3.set_string_value("v3");
+  HashedValue hv1(v1), hv2(v2), hv3(v3);
+
+  // Match with weighted cluster metadata and route action metadata.
+  {
+
+    RouteConstSharedPtr route = matcher.route(metadata, 0);
+    EXPECT_NE(nullptr, route);
+    EXPECT_NE(nullptr, route->routeEntry());
+
+    const Envoy::Router::MetadataMatchCriteria* criteria =
+        route->routeEntry()->metadataMatchCriteria();
+    EXPECT_NE(nullptr, criteria);
+    const std::vector<Envoy::Router::MetadataMatchCriterionConstSharedPtr>& mmc =
+        criteria->metadataMatchCriteria();
+    EXPECT_EQ(3, mmc.size());
+
+    EXPECT_EQ("k1", mmc[0]->name());
+    EXPECT_EQ(hv1, mmc[0]->value());
+
+    EXPECT_EQ("k2", mmc[1]->name());
+    EXPECT_EQ(hv2, mmc[1]->value());
+
+    EXPECT_EQ("k3", mmc[2]->name());
+    EXPECT_EQ(hv3, mmc[2]->value());
+  }
+
+  // Match with just route action metadata.
+  {
+    RouteConstSharedPtr route = matcher.route(metadata, 2001);
+    EXPECT_NE(nullptr, route);
+    EXPECT_NE(nullptr, route->routeEntry());
+    const Envoy::Router::MetadataMatchCriteria* criteria =
+        route->routeEntry()->metadataMatchCriteria();
+    EXPECT_NE(nullptr, criteria);
+    const std::vector<Envoy::Router::MetadataMatchCriterionConstSharedPtr>& mmc =
+        criteria->metadataMatchCriteria();
+    EXPECT_EQ(2, mmc.size());
+
+    EXPECT_EQ("k1", mmc[0]->name());
+    EXPECT_EQ(hv1, mmc[0]->value());
+
+    EXPECT_EQ("k2", mmc[1]->name());
+    EXPECT_EQ(hv2, mmc[1]->value());
+  }
+
+  // Match with weighted cluster metadata and route action metadata (merged).
```

Comment:
May be `'metadata_match' of both weighted cluster and route action are configured and with same metadata entry key.`